### ISSUE TITLE
Python code style conventions

### ISSUE THREAD

**Issue by jayvdb** [storyscript/community#16]
This is a general open discussion for ideas and plans for enforcement of Python code style conventions.

The story repo has a [`format` command](https://github.com/storyscript/cli/blob/f4888e395c616642441d060028e8fcc811ab0fc9/setup.py#L39) to use `black`, which is becoming the Python equivalent of `gofmt`, and black is being very rapidly adopted by the Python world.

https://black.readthedocs.io/en/stable/the_black_code_style.html provides the details.

On the flip side, flake8 has a very large catalog of plugins which do deep static analysis, providing automated QA that black can not do.

By default flake8 is incompatible with black, notably the line length.  These two tools can be used together by configuring flake8 to match black, with some unsolved corner cases.  Getting them working happily together can be infuriating at times, so having our cake and eating it will occasionally cost real time.

**Comment by jayvdb** [comment]
My recommendation is to enforce black and flake8 in CI across all Python repos, with black given priority, and flake8 config adjusted/disabled/whatever accordingly.

The benefit is developers know the rules, and there is very little personal decision making involved.  New flake8 plugins would get proposed in this repo and enforced on all repos, with local repo decision making for when to `noqa` rules where black and flake8 disagree.

**Comment by wilzbach** [comment]
> My recommendation is to enforce black and flake8 in CI across all Python repos, with black given priority, and flake8 config adjusted/disabled/whatever accordingly.

Sounds fair.

> Getting them working happily together can be infuriating at times, so having our cake and eating it will occasionally cost real time.

AFAICT it's just about disabling/changing flake8 checks that conflict?

So the only annoying part will be that we have to sync the configurations across multiple repositories.
However, how about maintaining them in a meta repository + a small update script (maybe even via automated bot like e.g. storyscript updates) that sends configuration updates to all Storyscript projects?

**Comment by jayvdb** [comment]
Note please do not jump the gun and make any large scale changes until a decision is done.  Also any large scale style changes should not be done by a human account - it should be a separate account, so that the effected lines are not re-attributed, and scripts which uses `git blame` can be taught the ignore the mass change.

Re deploying the config to all repo, I have previously suggested using `moban` at https://github.com/storyscript/community/issues/7 and I again recommend that approach.  I will implement it unless there are better tools selected and someone else it better suited to implement it.

**Comment by wilzbach** [comment]
>  Note please do not jump the gun and make any large scale changes until a decision is done.

I was talking about the configuration management.

> I again recommend that approach.

I was talking about automating config updates like we do here https://github.com/storyscript/runtime/pull/605.

So it's really just about copying a `flake8` config from the master repo around.
Individual projects can then do (if they every need to overwrite it) sth like:

```
flake8 --config .flake8/storyscript.config --append-config=project.config
```

> I have previously suggested using moban at #7 and I again recommend that approach.

So do I understand you correctly that your plan is to generate a project's `tox.ini` based on templates?

**Comment by jayvdb** [comment]
> I was talking about the configuration management.

Yes, so am I.  All of that will be done after collaborative decision making.

> So do I understand you correctly that your plan is to generate a project's tox.ini based on templates?

That isn't a priority, and is unrelated to code style (settngs shouldnt go in tox.ini; they should be in `setup.cfg` or `pyproject.toml`) but uniformity in the `tox.ini` will be something I will be working on, and templates may help with that.
